### PR TITLE
Perplexity

### PR DESCRIPTION
**Pull request by gideonite**
### Summary

Perplexity seems to work. I wonder if this is efficient enough since it redoes the loss calculation for each sentence. But, it doesn't really matter because it isn't part of the training computation. We can always just compute the perplexity on each epoch instead of on each batch... 

See commit messages for more details. There's some other stuff in here, including caching the preprocessing step in `/tmp`.

### CODE REVIEW

File: utils.py [hunk 1]
```diff
@@ -70,26 +71,39 @@ class DataReader(object):
...
-        self.vocab.construct(path, vocab_size)
+        if os.path.isfile(self.cache_file):
+            print('loading cached data from: %s' % self.cache_file)
+            with open(self.cache_file, 'rb') as f:
+                self.data = pickle.load(f)
+        else:
+            print('constructing data set')
+
+            # intial pass over path to construct vocabulary 
+            self.vocab.construct(path, vocab_size)
```
Comment (stefangstark) [inline utils.py:88]: I think you either need to save & load the vocabulary or move this outside the if/else

File: lstm.py [hunk 2]
```diff
@@ -101,18 +115,41 @@ def add_model(self, input_data):
         return sentence_logits
 
 
+    def add_perplexity_op(self, sentence_logits):
+        # TODO is this redundant with add_loss_op? Maybe be incurring
```
Comment (stefangstark) [inline lstm.py:119]: This seems a little unsatisfying... but I think that the log sum means we can't just make this a function of a scalar loss

Comment (gideonite) [reply lstm.py:119]: Right. That's what I thought too but I didn't go through the exercise of writing out the equations. It may work out just fine but I figured to just run with this for now.

File: utils.py (hunk 1 above)
Comment (gideonite) [reply utils.py:88]: Fixed by b9b2954f042855a777030b43178dacb5b92f0efc

Where does `vocab` actually get used?

Comment (stefangstark) [reply utils.py:88]: The model only needs the tokens, but its nice to keep the words around for interpretation. We might need it for the sentence generation
